File: 3620-maximum-number-of-distinct-elements-after-operations/maximum-number-of-distinct-elements-after-operations.py
class Solution:
    def maxDistinctElements(self, nums: List[int], k: int) -> int:
        nums.sort()
        n = len(nums)

        prev, count = nums[0] - k, 1

        for i in range(1, n):
            minval = nums[i] - k
            maxval = nums[i] + k

            if prev < minval:
                prev = minval
            elif prev < maxval:
                prev += 1
            else:
                continue

            count += 1

        return count

File: 3620-maximum-number-of-distinct-elements-after-operations/maximum-number-of-distinct-elements-after-operations.cpp
#define ll long long
#define vi vector<int>
#define vll vector<ll>
#define vs vector<string>
#define mpi map<int, int>
#define mpll map<ll, ll>
#define mps map<string, int>
#define umpi unordered_map<int, int>
#define umpll unordered_map<ll, ll>
#define pb push_back
#define all(x) (x).begin(), (x).end()

const int MOD = 1e9 + 7;

class Solution {
public:
    int maxDistinctElements(vector<int>& nums, int k) {
        sort(all(nums));
        ll cur = LLONG_MIN, count = 0;

        for (auto it : nums) {
            ll left = it - k;
            ll right = it + k;
            if (cur < left) {
                cur = left;
            }
            if (cur <= right) {
                count++;
                cur++;
            }
        }

        return count;
    }
};